internal/app/go-filecoin/plumbing/msg/waiter.go

```diff
@@ -149,149 +134,94 @@ func (w *Waiter) waitForMessage(ctx context.Context, ch <-chan interface{}, msgC
-	if err != nil {
-		return nil, err
-	}
+		for i, msg := range blkMsgs {
+			if cids[i].Equals(msgCid) {
+				// wrapped cid might be different from given cid
+				targetCid, err := msg.Cid()
+				if err != nil {
+					return nil, false, err
+				}
 
-	var tsMessages [][]*types.SignedMessage
-	for i := 0; i < ts.Len(); i++ {
-		blk := ts.At(i)
-		secpMsgs, _, err := w.messageProvider.LoadMessages(ctx, blk.Messages)
-		if err != nil {
-			return nil, err
+				recpt, err := w.receiptByIndex(ctx, ts.Key(), targetCid, tsMessages)
+				if err != nil {
+					return nil, false, errors.Wrap(err, "error retrieving receipt from tipset")
+				}
+				return &ChainMessage{msg, blk, recpt}, true, nil
+			}
 		}
-		tsMessages = append(tsMessages, secpMsgs)
 	}
+	return nil, false, nil
+}
 
-	res, err := consensus.NewDefaultProcessor().ProcessTipSet(ctx, st, vm.NewStorageMap(w.bs), ts, tsMessages, ancestors)
+func (w *Waiter) receiptByIndex(ctx context.Context, tsKey block.TipSetKey, targetCid cid.Cid, messages [][]*types.SignedMessage) (*types.MessageReceipt, error) {
+	receiptCid, err := w.chainReader.GetTipSetReceiptsRoot(tsKey)
 	if err != nil {
 		return nil, err
 	}
 
-	// If this is a failing conflict message there is no application receipt.
-	_, failed := res.Failures[msgCid]
-	if failed {
-		return nil, nil
+	receipts, err := w.messageProvider.LoadReceipts(ctx, receiptCid)
+	if err != nil {
+		return nil, err
 	}
 
-	j, err := w.msgIndexOfTipSet(ctx, msgCid, ts, res.Failures)
+	deduped, err := consensus.DeduppedMessages(messages)
 	if err != nil {
 		return nil, err
 	}
-	// TODO #3194: out of bounds receipt index should return an error.
-	if j < len(res.Results) {
-		rcpt = res.Results[j].Receipt
-	}
-	return rcpt, nil
-}
 
-// msgIndexOfTipSet returns the order in which msgCid appears in the canonical
-// message ordering of the given tipset, or an error if it is not in the
-// tipset.
-// TODO: find a better home for this method
-func (w *Waiter) msgIndexOfTipSet(ctx context.Context, msgCid cid.Cid, ts block.TipSet, fails map[cid.Cid]struct{}) (int, error) {
-	duplicates := make(map[cid.Cid]struct{})
-	var msgCnt int
-	for i := 0; i < ts.Len(); i++ {
-		secpMsgs, _, err := w.messageProvider.LoadMessages(ctx, ts.At(i).Messages)
-		if err != nil {
-			return -1, err
-		}
-		for _, msg := range secpMsgs {
-			c, err := msg.Cid()
+	receiptIndex := 0
+	for _, blkMessages := range deduped {
+		for _, msg := range blkMessages {
+			msgCid, err := msg.Cid()
 			if err != nil {
-				return -1, err
-			}
-			_, failed := fails[c]
-			if failed {
-				continue
-			}
-			_, isDup := duplicates[c]
-			if isDup {
-				continue
+				return nil, err
 			}
-			duplicates[c] = struct{}{}
-			if c.Equals(msgCid) {
-				return msgCnt, nil
+
+			if msgCid.Equals(targetCid) {
+				if receiptIndex >= len(receipts) {
+					return nil, errors.Errorf("could not find message receipt at index %d", receiptIndex)
```

Comment:
I think this only gets hit if a programmer is violating system invariants and that this error should shout louder.  The only way this can happen is if there is an inconsistency in the stored receipts between initial dedupping and the waiter's deduping.

Reply:
This is true, unless there is some other sort of corruption. I've been violating these system invariants often enough while programming this that I still feel the check is worth it.